File: app/src/main/java/berlindroid/zethree/UpdateActivity.java
package berlindroid.zethree;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.Nullable;
import berlindroid.zethree.util.LowQualityUpdateDownloader;
import berlindroid.zethree.util.LowQualityUpdateDownloader.ResolvedApkUrl;

public class UpdateActivity extends Activity {

    @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.update_activity);
    }

    // risky only if you want it to be risky !!!!!!111one
    @SuppressLint("StaticFieldLeak")
    @Override protected void onStart() {
        super.onStart();

        // AsyncTask is the new Coroutine Job!

        //noinspection deprecation
        new AsyncTask<Void, Void, Void>() {
            @Override protected Void doInBackground(Void... voids) {
                // cause some fake delay to prevent rate-limiter from kicking in
                try { Thread.sleep(5000); } catch (InterruptedException ignored) {}
                new Handler(Looper.getMainLooper()).post(
                    () -> findViewById(R.id.updateWaitText).setVisibility(View.VISIBLE)
                );
                return null;
            }
        }.execute();

        //noinspection deprecation
        new AsyncTask<Void, Void, Void>() {
            @Override protected Void doInBackground(Void... voids) {
                LowQualityUpdateDownloader downloader =
                    new LowQualityUpdateDownloader(UpdateActivity.this);
                ResolvedApkUrl latestApkData = downloader.getLatestApkUrl(
                    "https://api.github.com/repos/gdg-berlin-android/ZeThree/releases"
                );

                new Handler(Looper.getMainLooper()).post(
                    () -> {
                        if (latestApkData == null) {
                            Toast.makeText(
                                UpdateActivity.this,
                                "Can't fetch the update",
                                Toast.LENGTH_LONG
                            ).show();
                        } else if (BuildConfig.VERSION_NAME.equals(latestApkData.appVersion)) {
                            Toast.makeText(
                                UpdateActivity.this,
                                "Already at the latest version",
                                Toast.LENGTH_LONG
                            ).show();
                        } else {
                            Toast.makeText(
                                UpdateActivity.this,
                                "Install the downloaded file",
                                Toast.LENGTH_LONG
                            ).show();

                            Intent i = new Intent(Intent.ACTION_VIEW);
                            i.setData(Uri.parse(latestApkData.url));
                            startActivity(i);
                        }

                        UpdateActivity.this.finish();
                    }
                );
                return null;
            }
        }.execute();
    }

}

File: app/src/main/java/berlindroid/zethree/UpdateActivity.kt
package berlindroid.zethree

import android.app.Activity
import android.os.Bundle
import android.annotation.SuppressLint
import android.os.AsyncTask
import android.os.Looper
import berlindroid.zethree.util.LowQualityUpdateDownloader
import android.widget.Toast
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.view.View
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.coroutineScope
import kotlin.coroutines.coroutineContext

class UpdateActivity : Activity() {

    //Use this scope to run the coroutine and remove the AsyncTasks :*
    private val coroutineScope = Dispatchers.IO + Job()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.update_activity)
    }

    // risky only if you want it to be risky !!!!!!111one
    @SuppressLint("StaticFieldLeak")
    override fun onStart() {
        super.onStart()


                // AsyncTask is the new Coroutine Job!
        object : AsyncTask<Void?, Void?, Void?>() {
            override fun doInBackground(vararg params: Void?): Void? {
                // cause some fake delay to prevent GitHub's rate-limiter from kicking in
                try {
                    Thread.sleep(1000)
                } catch (ignored: InterruptedException) {
                }
                Handler(Looper.getMainLooper()).post {
                    findViewById<View>(R.id.updateWaitText).visibility = View.VISIBLE
                }
                return null
            }
        }.execute()
        object : AsyncTask<Void?, Void?, Void?>() {
            override fun doInBackground(vararg params: Void?): Void? {
                val result = LowQualityUpdateDownloader.resolveLatestApk()
                Handler(Looper.getMainLooper()).post {
                    when (result.status) {
                        LowQualityUpdateDownloader.Status.FAILED -> Toast.makeText(
                            this@UpdateActivity,
                            "Can't fetch the update: " + result.errorMessage,
                            Toast.LENGTH_LONG
                        ).show()
                        LowQualityUpdateDownloader.Status.APK_AVAILABLE -> {
                            Toast.makeText(
                                this@UpdateActivity,
                                "Next: Install the downloaded file",
                                Toast.LENGTH_LONG
                            ).show()
                            val i = Intent(Intent.ACTION_VIEW)
                            i.data = Uri.parse(result.url)
                            startActivity(i)
                        }
                        LowQualityUpdateDownloader.Status.ALREADY_LATEST -> Toast.makeText(
                            this@UpdateActivity,
                            "Already at the latest version: v" + result.appVersion,
                            Toast.LENGTH_LONG
                        ).show()
                    }
                    finish()
                }
                return null
            }
        }.execute()
    }
}